graphdatascience/aura_sessions.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import List, Optional, Tuple
+
+from neo4j import GraphDatabase
+
+from graphdatascience import GraphDataScience
+from graphdatascience.aura_api import AuraApi, InstanceDetails
+from graphdatascience.query_runner.aura_db_arrow_query_runner import (
+    AuraDbConnectionInfo,
+)
+
+
+@dataclass
+class SessionInfo:
+    name: str
+
+
+class AuraSessions:
+    GDS_SESSION_NAME_PREFIX = "gds-session-"
+
+    def __init__(
+        self,
+        db_credentials: AuraDbConnectionInfo,
+        aura_api_client_auth: Tuple[str, str],
+        tenant_id: Optional[str] = None,
+    ) -> None:
+        self._db_credentials = db_credentials
+        self._aura_api = AuraApi(
+            tenant_id=tenant_id, client_id=aura_api_client_auth[0], client_secret=aura_api_client_auth[1]
+        )
+
+    # TODO wrapper around it for closable?
+    def create_gds(self, session_name: str, session_password: str) -> GraphDataScience:
+        if len(session_password) < 8:
+            raise ValueError("Password must be at least 8 characters long.")
+
+        if session_name in [session.name for session in self.list_sessions()]:
+            raise RuntimeError(f"Session with name `{session_name}` already exists.")
+
+        create_details = self._aura_api.create_instance(AuraSessions._instance_name(session_name))
+        self._aura_api.wait_for_instance_running(create_details.id)
+
+        gds_user = create_details.username
+        url = create_details.connection_url
+
+        self._change_initial_pw(
+            gds_url=url, gds_user=gds_user, initial_pw=create_details.password, new_pw=session_password
+        )
+
+        return self._construct_client(gds_url=url, gds_user=gds_user, gds_pw=session_password)
+
+    def connect(self, session_name: str, session_password: str) -> GraphDataScience:
+        instance_name = AuraSessions._instance_name(session_name)
+        matched_instances = [
+            instance.id for instance in self._aura_api.list_instances() if instance.name == instance_name
+        ]
+
+        if len(matched_instances) != 1:
+            raise ValueError("TODO")
+
+        instance_details = self._aura_api.list_instance(matched_instances[0])
+
+        if instance_details:
+            gds_url = instance_details.connection_url
+        else:
+            raise RuntimeError(
+                f"Unable to get connection information for session `{session_name}`. Does it still exist?"
+            )
+
+        # Hardcoded neo4j user as sessions are always created with this user
```

Comment:
Why is that the case? Maybe it would be good if it was the gds_user and we would check immediately if the user is elegible of querying the database to throw early if it lacks permissions.

Reply:
The reason is that we are creating a new DBMS instance, which always comes with one default user called `neo4j`. We are setting the password to our user's input password (for reconnecting), but I don't see why we should also allow them to specify a username. The session name serves the name-identity function now. Sure, they _can_ log in and do stuff like `CREATE USER`, but do we really need to support that when we are wanting to remove the DBMS, including system db and its RBAC?

Reply:
Indeed this is the `gds_user` and only used for logging into the GDS session (AuraDS instance).

Reply:
I was actually thinking about the username we issue the arrow requests from, but I guess that one is retrieved from the token which we acquire from the actual username that is used in AuraDB. So it should be fine with the default usename here